nx/lib/io/nx_io.ml

```diff
@@ -100,6 +100,162 @@ module Safe = struct
   let save_npz ?(overwrite = true) path items =
     Nx_npy.save_npz ~overwrite path items
 
+  (* Text I/O *)
+  let save_txt ?(sep = " ") ?(append = false) ~out (type a) (type b)
+      (arr : (a, b) Nx.t) =
+    try
+      let rank = Array.length (Nx.shape arr) in
+      if rank <> 1 && rank <> 2 then Error Unsupported_shape
+      else
+        let perm = 0o666 in
+        let open_flags =
+          if append then [ Open_wronly; Open_creat; Open_append; Open_text ]
+          else [ Open_wronly; Open_creat; Open_trunc; Open_text ]
+        in
+        let oc = open_out_gen open_flags perm out in
+        Fun.protect
+          ~finally:(fun () -> close_out oc)
+          (fun () ->
+            let shape = Nx.shape arr in
+            let rows, cols =
+              match shape with
+              | [| n |] -> (1, n)
+              | [| r; c |] -> (r, c)
+              | _ -> assert false
+            in
+            let arr2d_any = Nx.reshape [| rows; cols |] arr in
+            let res =
+              match Nx.dtype arr with
+              | Nx_core.Dtype.Float32 | Nx_core.Dtype.Float64 ->
+                  let arrf : (float, Nx.float64_elt) Nx.t =
+                    Nx.astype Nx.float64 arr2d_any
+                  in
+                  for i = 0 to rows - 1 do
+                    for j = 0 to cols - 1 do
+                      if j > 0 then output_string oc sep;
+                      Printf.fprintf oc "%g" (Nx.item [ i; j ] arrf)
+                    done;
+                    output_char oc '\n'
+                  done;
+                  Ok ()
+              | Nx_core.Dtype.Int32 ->
+                  let arri32 : (int32, Nx.int32_elt) Nx.t =
+                    Nx.astype Nx.int32 arr2d_any
+                  in
+                  for i = 0 to rows - 1 do
+                    for j = 0 to cols - 1 do
+                      if j > 0 then output_string oc sep;
+                      output_string oc
+                        (Int32.to_string (Nx.item [ i; j ] arri32))
+                    done;
+                    output_char oc '\n'
+                  done;
+                  Ok ()
+              | Nx_core.Dtype.Int64 ->
+                  let arri64 : (int64, Nx.int64_elt) Nx.t =
+                    Nx.astype Nx.int64 arr2d_any
+                  in
+                  for i = 0 to rows - 1 do
+                    for j = 0 to cols - 1 do
+                      if j > 0 then output_string oc sep;
+                      output_string oc
+                        (Int64.to_string (Nx.item [ i; j ] arri64))
+                    done;
+                    output_char oc '\n'
+                  done;
+                  Ok ()
+              | _ -> Error Unsupported_dtype
```

Comment:
We should support all the dtypes or at least a sensible subset of what NumPy supports, including bool